src/functions.cpp

```diff
@@ -0,0 +1,169 @@
+#include "functions.h"
+
+#include "log.h"
+
+namespace bpftrace {
+
+namespace {
+std::string arg_types_str(const std::vector<SizedType> &arg_types)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const SizedType &arg_type : arg_types) {
+    if (!first)
+      str += ", ";
+    str += typestr(arg_type);
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+
+std::string param_types_str(const std::vector<Param> &params)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const Param &param : params) {
+    if (!first)
+      str += ", ";
+    str += typestr(param.type());
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+} // namespace
+
+FunctionRegistry::FunctionRegistry(std::ostream &out) : out_(out)
+{
+}
+
+const Function &FunctionRegistry::add(Function::Origin origin,
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  return add(origin, {}, name, return_type, params);
+}
+
+const Function &FunctionRegistry::add(Function::Origin origin,
+                                      std::string_view ns,
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  FqName fq_name{
+    .ns = std::string{ ns },
+    .name = std::string{ name },
+  };
+
+  // Check for duplicate function definitions
+  // The assumption is that builtin functions are all added to the registry
+  // before any user-defined functions.
+  // Builtin functions can be duplicated. Other functions can not.
+  for (const Function &func : funcs_by_fq_name_[fq_name]) {
+    if (func.origin() != Function::Origin::Builtin) {
+      throw std::runtime_error("Cannot redefine existing function: " +
+                               fq_name.str() + param_types_str(params));
```

Comment:
Feels a bit odd to do policy check here. Seems like it should belong in caller.  But assuming we really want to keep it here, I think you should write the error to `out_` and return a nullptr. Reference type return means infallability - seems like you have a very possible, not abnormal failure possible here (eg user defines two `foo()`s).

Reply:
There'll be multiple callers (at least one for each of builtin, UDF and external functions), so I think it makes sense to centralise the logic here to ensure `FunctionRegistry` can never contain duplicates.